code/environments/stanford_client.py

```diff
@@ -0,0 +1,523 @@
+
+            trap1_y = np.random.rand() * (self.dark_line - self.init_strip_y[1]) + self.init_strip_y[1]
+            trap2_y = np.random.rand() * (self.dark_line - self.init_strip_y[1]) + self.init_strip_y[1]
+            self.test_trap_y = [[trap1_y, trap1_y+trap_size], [trap2_y, trap2_y+trap_size]]
+
+            print(self.test_trap_x, self.test_trap_y, "\n\n\n")
+
+        # if random_obs:
+        #     obs = self.observation_space.sample()
+        # else:
+        normalization_data = self.preprocess_data()
+        obs = self.get_observation(normalization_data=normalization_data)
+        return obs
+
+    def step(self, action, random_obs=False, action_is_vector=False):
+        # random_obs = True only for debugging purposes
+        episode_length = sep.max_steps
+        curr_state = self.state
+
+        # Get the observation at the current state to provide PlaNet the expected output
+        # if random_obs:
+        #     obs = self.observation_space.sample()
+        # else:
+        normalization_data = self.preprocess_data()
+        obs = self.get_observation(normalization_data=normalization_data)
+        # obs = OrderedDict()        
+        # obs['image'] = obs_nav
+
+        self._step += 1
+        
+        if action_is_vector:
+            # Normalizing actions
+            if np.linalg.norm(action) > 0.2:
+                action = action / np.linalg.norm(action) * 0.2
+            
+            new_theta = np.arctan2(action[1], action[0])
+            if new_theta < 0:  # Arctan stuff
+                new_theta += 2*np.pi
+            next_state = curr_state + action
+        else:
+            new_theta = action * np.pi + np.pi
+            vector = np.array([np.cos(new_theta), np.sin(new_theta)]) * sep.velocity  # Go in the direction the new theta is
+            next_state = curr_state + vector
+    
+        cond_hit = self.detect_collision(next_state)
+
+        # Previous value of reached_goal 
+        temp_reached_goal = self.reached_goal
+
+        # If already reached goal, don't move
+        if not temp_reached_goal:
+            if self.in_goal(next_state):
+                self.state = next_state
+                self.orientation = new_theta
+                self.reached_goal = True
+                #self.done = True
+            elif cond_hit == False:  # If collided, don't move. Else move.
+                self.state = next_state
+                self.orientation = new_theta
+
+        # If we just reached the goal, collect reward
+        # But if we already reached the goal previously, get reward 0
+        if not temp_reached_goal and self.reached_goal:
+            reward = sep.epi_reward 
+        else:
+            reward = 0
+        
+        # If already reached goal, don't reason about trap rewards
+        if not temp_reached_goal:
+            cond_false = self.in_trap(next_state)
+            reward -= sep.epi_reward * cond_false
+
+        self.done = self._step >= episode_length - 1

```

Comment:
Should modify how self.done and self.reached_goal are treated with respect to how DVRL handles the agent reaching the goal.